### ISSUE TITLE
[MNG-7395] Support interpolation in extensions.xml

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#9171]
**[Christoph Läubrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=laeubi)** opened **[MNG-7395](https://issues.apache.org/jira/browse/MNG-7395?redirect=false)** and commented

Currently an entry has to be the following form:

```xml
<extension>
	<groupId>org.eclipse.tycho</groupId>
	<artifactId>tycho-build</artifactId>
	<version>2.7.0-SNAPSHOT</version>
</extension>
```

This has the drawback that I always need to edit the extension.xml file and there is no way to specify it on the commandline.

The proposal is to allow the follwoing:

```xml
<extension>
	<groupId>org.eclipse.tycho</groupId>
	<artifactId>tycho-build</artifactId>
	<version>${tycho-version|2.6.0}</version>
</extension>
```

I would then expect the following:
1. If no systemproperty with name 'tycho-version' exits, the value after the | is used as a default
2. If a systemproperty with name 'tycho-version' exits it is used as a version

That way I can call mvn -Dtycho-version=2.7.0-SNAPSHOT clean install to override the default version.

---

**Remote Links:**
- [GitHub Pull Request #136
](https://github.com/apache/maven-integration-testing/pull/136)
- [GitHub Pull Request #665
](https://github.com/apache/maven/pull/665)
- [GitHub Pull Request #673
](https://github.com/apache/maven/pull/673)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.5](https://github.com/apache/maven/milestone/106?closed=1)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You are conflating two issue here actually:
* Interpolation of `extensions.xml`
* Property default values with wellknown `value:-default` approach

Please split your ticket.

**Comment by jira-importer** [comment]
**[Christoph Läubrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=laeubi)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Honestly I don't expect full blown interpolation or generic default value handling.
As extensions.xml is read at a very low level, I just seeking for a simple approach to make it more flexible, that is backward compatible, and could happen before maven-9 ;-)

I have made a local patch I can share to demonstrate a simple solution for this and if you like I can share it to discuss this further.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

We'd need at least this cheap solution: https://github.com/apache/maven/blob/master/maven-settings-builder/src/main/java/org/apache/maven/settings/building/DefaultSettingsBuilder.java#L113

**Comment by jira-importer** [comment]
**[Christoph Läubrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=laeubi)** commented

If we can have more support, its even better I'm just not sure if something like that is available at this stage.

I have created a draft PR here: https://github.com/apache/maven/pull/665/files so if you can take a look and give some insight that would be great.

**Comment by jira-importer** [comment]
**[Christoph Läubrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=laeubi)** commented

I think this is ready now and I'd like to backport this to maven 3.8.x as well so how could we proceed here?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [2798ee91968b31aa444cfd7ded14ada0a9f043f4](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=2798ee91968b31aa444cfd7ded14ada0a9f043f4), with [a5acd3ec6095ea2ff50191576c33685c1d038c1f](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=a5acd3ec6095ea2ff50191576c33685c1d038c1f) for `maven-3.8.x` branch and [f26315a3809d01a229db5fe437bb7dd07c899c35](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=f26315a3809d01a229db5fe437bb7dd07c899c35).

**Comment by jira-importer** [comment]
**[Christoph Läubrich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=laeubi)** commented

Thanks! Can't await to see this in maven release :-)
